File: guests/java/dev/kaya/milestone2kt/Panes.java
package dev.kaya.milestone2kt;

import dev.kaya.KayaApp;

/**
 * The panes conformance scene from the JVM — a THREE-pane ceiling as
 * assertions (docs/multicolumn-plan.md D1/D5). Nothing here is
 * panes-specific except {@code panes(3)}, asked for ONCE; the stack is
 * the ordinary navigation stack. See guests/rust/panes.rs and
 * tools/scenes/panes.steps.
 */
final class Panes {
    private static final long CONTENT = 7;
    private static final long DETAIL = 8;

    static void app() {
        KayaApp app = new KayaApp();

        app.build(tx -> {
            tx.window(0).title("panes").panes(3);
            KayaApp.Signal<String> root = tx.signal("root pane");
            tx.mount(tx.column(() -> {
                tx.label(root).a11yId("root"); // label#0
                tx.button("open content", inner -> { // button#0
                    long entry = inner.pushEntry(CONTENT).title("content").id();
                    KayaApp.Widget pane = inner.column(() -> {
                        KayaApp.Signal<String> caption = inner.signal("content pane");
                        inner.label(caption).a11yId("content"); // label#1
                        inner.button("open detail", deep -> { // button#1
                            long leaf = deep.pushEntry(DETAIL).title("detail").id();
                            KayaApp.Widget detail = deep.column(() -> {
                                KayaApp.Signal<String> tail = deep.signal("detail pane");
                                deep.label(tail).a11yId("detail"); // label#last
                            });
                            deep.mountIn(leaf, detail);
                        });
                    });
                    inner.mountIn(entry, pane);
                });
            }));
        });

        app.dispatchLoop();
    }

    private Panes() {}
}

File: guests/go/panes/panes.go
// The panes scene (tools/scenes/panes.steps): nothing here is
// panes-specific except `Panes(3)`, asked for ONCE.
package panes

import (
	kaya "dev.kaya/bindings/go"
)

const (
	content = 7
	detail  = 8
)

func App() *kaya.App {
	app := kaya.NewApp()

	app.Build(func(tx *kaya.Tx) {
		tx.Window(0).Title("panes").Panes(3)

		tx.Mount(tx.Column(func() {
			// Authored ids: an index read passes for an empty arm.
			caption := tx.Signal("root pane")
			tx.Label(caption).A11yID("root")       // label#0
			tx.Button("open content", openContent) // button#0
		}))
	})

	return app
}

func openContent(tx *kaya.Tx) {
	entry := tx.PushEntry(content).Title("content").Id()
	pane := tx.Column(func() {
		caption := tx.Signal("content pane")
		tx.Label(caption).A11yID("content")  // label#1
		tx.Button("open detail", openDetail) // button#1
	})
	tx.MountIn(entry, pane)
}

func openDetail(tx *kaya.Tx) {
	entry := tx.PushEntry(detail).Title("detail").Id()
	pane := tx.Column(func() {
		caption := tx.Signal("detail pane")
		tx.Label(caption).A11yID("detail") // label#last
	})
	tx.MountIn(entry, pane)
}
